File: numeric_operation/hanoi.rb
# ハノイの塔
#
# 3本のポールとサイズの異なるn個のドーナツ円盤がある
# 出発点のポールに固まっている円盤を全て目的のポールに移す。

POLES = ["A","B","C"]

def hanoi(n, from, to)
  return if n == 0
  temp = POLES - [from,to].to_s
  hanoi(n-1, from, temp)
  puts "move disk##{n} from #{from} to #{to}"
  hanoi(n-1, temp, to)
  return
end

hanoi(4,POLES.first,POLES.last)

File: numeric_operation/hanoi.py
#!/usr/bin/env python
# -*- coding: utf-8


def hanoi(n, from_stack, to_stack, tmp_stack):
    if n <= 1:
        print(f'move {n} to disk from {from_stack} to {to_stack}')
        return
    # 最後のディスクの移動のため、他ディスクを別場所に退避させる
    hanoi(n-1, from_stack, tmp_stack, to_stack)
    # 最後のディスクの移動
    print(f'move {n} to disk from {from_stack} to {to_stack}')
    # 最後のディスクの上に、退避させておいた他ディスクを移動する
    hanoi(n-1, tmp_stack, to_stack, from_stack)


if __name__ == '__main__':
    n = int(input("n: "))
    from_stack = input("from: ")
    to_stack = input("to: ")
    tmp_stack = input("tmp: ")
    hanoi(n, from_stack, to_stack, tmp_stack)
